cmd/ibctest/ibc_test.go

```diff
@@ -0,0 +1,168 @@
+
+		testMatrix.Relayers = []string{"rly"}
+		testMatrix.ChainSets = [][]ibc.BuiltinChainFactoryEntry{
+			{
+				{Name: "gaia", Version: "v6.0.4", ChainID: "cosmoshub-1004", NumValidators: 2, NumFullNodes: 1},
+				{Name: "osmosis", Version: "v7.0.4", ChainID: "osmosis-1001", NumValidators: 2, NumFullNodes: 1},
+			},
+		}
+
+		return nil
+	}
+
+	// Otherwise parse the given file.
+	fmt.Fprintf(os.Stderr, "Loading matrix file from %s\n", mainFlags.MatrixFile)
+	j, err := os.ReadFile(mainFlags.MatrixFile)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(j, &testMatrix); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateTestMatrix() error {
+	for _, r := range testMatrix.Relayers {
+		if _, err := getRelayerFactory(r); err != nil {
+			return err
+		}
+	}
+
+	for _, cs := range testMatrix.ChainSets {
+		if _, err := getChainFactory(cs); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func getRelayerFactory(name string) (ibc.RelayerFactory, error) {
+	switch name {
+	case "rly", "cosmos/relayer":
+		return ibc.NewBuiltinRelayerFactory(ibc.CosmosRly), nil
+	case "hermes":
+		return ibc.NewBuiltinRelayerFactory(ibc.Hermes), nil
+	default:
+		return nil, fmt.Errorf("unknown relayer type %q (valid types: rly, hermes)", name)
+	}
+}
+
+func getChainFactory(chainSet []ibc.BuiltinChainFactoryEntry) (ibc.ChainFactory, error) {
+	if len(chainSet) != 2 {
+		return nil, fmt.Errorf("chain sets must have length 2 (found a chain set of length %d)", len(chainSet))
+	}
+
+	return ibc.NewBuiltinChainFactory(chainSet), nil
+}
+
+// TestRelayer is the root test for the relayer.
+// It runs each subtest in parallel;
+// if this is too taxing on a system, the -test.parallel flag
+// can be used to reduce how many tests actively run at once.
+func TestRelayer(t *testing.T) {
+	t.Parallel()
+
+	// One layer of subtests for each relayer to be tested.
+	for _, r := range testMatrix.Relayers {
+		rf, err := getRelayerFactory(r)
+		if err != nil {
+			// This error should have been validated before running tests.
+			panic(err)
+		}
+
+		t.Run(r, func(t *testing.T) {
+			t.Parallel()
+
+			// And another layer of subtests for each chainset.
+			for _, cs := range testMatrix.ChainSets {
+				cf, err := getChainFactory(cs)
+				if err != nil {
+					panic(err)
+				}
+
+				chainNames := make([]string, len(cs))
+				for i, c := range cs {
+					// TODO: this shouldn't need to truncate bits of the name and version,
+					// but without this, the tests hang on a failed container running gentx, with this error:
+					//    panic: failed to execute message; message index: 0: invalid moniker length; got: 75, max: 70
+					// It would be better to figure out the correct gentx invocation
+					// rather than hiding details from the test name.
+					chainNames[i] = c.Name[:3] + "@" + c.Version[1:]
+				}
+				chainTestName := strings.Join(chainNames, "+")
```

Comment:
with only 2 chains this should be fine, but I have noticed the docker containers can fail to run if the testname is too long since it is used in the hostname of the containers (max length 63), so we may want to truncate this value and add some other randomized string